### PR TITLE
Toy mixture model example

### PR DESCRIPTION
**Pull request by garrett-bernstein**
Adds a script for learning a toy mixture model to provide a simple example for discrete enumeration, as [suggested by Martin](https://forum.pyro.ai/t/summing-out-latent-discrete-parameters-in-mixture-model/1896/4?u=gbernstein).

### CODE REVIEW

**Comment by martinjankowiak** [comment]
great thanks @garrett-bernstein !!

could you please add a line to `tests/test_examples.py` for `toy_mixture_model.py` so that the tests pass?

**Comment by garrett-bernstein** [comment]
> could you please add a line to `tests/test_examples.py` for `toy_mixture_model.py` so that the tests pass?

Done. Also added a brief note to `CONTRIBUTIONS.md` with that instruction.

File: examples/toy_mixture_model.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import pyro
+import pyro.distributions as dist
+import pyro.infer
+import pyro.optim
+from pyro.ops.indexing import Vindex
+
+smoke_test = ('CI' in os.environ)
+
+
+def main():
```
Comment (martinjankowiak) [inline examples/toy_mixture_model.py:42]: nit: could you maybe remove some of the extra blank lines throughout? this is really just a stylistics thing, but we don't use so much whitespace in the rest of the codebase...

File: examples/toy_mixture_model.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    p_A = pyro.sample('p_A', dist.Beta(1, 1))
+
+    p_B = pyro.sample('p_B', dist.Beta(torch.ones(2), torch.ones(2)).to_event(1))
+
+    p_C = pyro.sample('p_C', dist.Beta(torch.ones(2), torch.ones(2)).to_event(1))
+
+    with pyro.plate('data_plate', n):
+        A = pyro.sample('A', dist.Bernoulli(p_A.expand(n)), obs=obs['A'])
+
+        B = pyro.sample('B', dist.Bernoulli(Vindex(p_B)[A.type(torch.long)]), infer={"enumerate": "parallel"})
```
Comment (martinjankowiak) [inline examples/toy_mixture_model.py:90]: could you maybe add a comment on the role of `Vindex`?

File: examples/toy_mixture_model.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    # setup svi object
+    svi = pyro.infer.SVI(model,
+                         guide,
+                         pyro.optim.Adam({}),
+                         loss=loss_func
+                         )
+
+    # perform svi
+    num_steps = 20000 if not smoke_test else 1
```
Comment (martinjankowiak) [inline examples/toy_mixture_model.py:122]: does it really need so many steps to converge? maybe you can change the learning rate to something like `0.01` and lower the number of steps?

Comment (martinjankowiak) [reply examples/toy_mixture_model.py:122]: e.g. i seem to get reasonable results with `lr=0.01` and 4000 steps

File: examples/toy_mixture_model.py (hunk 1 above)
Comment (garrett-bernstein) [reply examples/toy_mixture_model.py:42]: Done. Removed all the ones within methods. Left the double newlines between methods as per my PEP8 checker.

File: examples/toy_mixture_model.py (hunk 2 above)
Comment (garrett-bernstein) [reply examples/toy_mixture_model.py:90]: Done.

File: examples/toy_mixture_model.py (hunk 3 above)
Comment (garrett-bernstein) [reply examples/toy_mixture_model.py:122]: Done.

**Comment by garrett-bernstein** [comment]
> add files to `tutorials/source`. e.g. see `tutorial/source/sparse_gamma.rst` and the corresponding entry in `tutorial/source/index.rst`

Done.

**Comment by garrett-bernstein** [comment]
Whoops, looks like I "dismissed Martin's stale review". Sorry, haven't done PR via Github before, just Bitbucket...

File: tests/test_examples.py [hunk 4]
```diff
@@ -97,6 +97,7 @@
     'sparse_gamma_def.py --num-epochs=2 --eval-particles=2 --eval-frequency=1 --guide custom',
     'sparse_gamma_def.py --num-epochs=2 --eval-particles=2 --eval-frequency=1 --guide auto',
     'sparse_gamma_def.py --num-epochs=2 --eval-particles=2 --eval-frequency=1 --guide easy',
+    'toy_mixture_model.py',
```
Comment (fritzo) [inline tests/test_examples.py:100]:
```py
'toy_mixture_model.py --num-steps=1',
```

File: examples/toy_mixture_model.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+from torch.distributions.bernoulli import Bernoulli
+from torch.distributions.beta import Beta
+from tqdm import tqdm
+import pyro
+import pyro.distributions as dist
+import pyro.infer
+import pyro.optim
+from pyro.ops.indexing import Vindex
+
+smoke_test = ('CI' in os.environ)
```
Comment (fritzo) [inline examples/toy_mixture_model.py:37]: We use the environment variable idiom `'CI' in os.environ` idiom only for notebook tutorials where it is difficult to pass command line arguments. For script tutorials we instead use `argparse` and pass options to the script test in `test_examples.py`. Could you remove the `smoke_test = ...` and instead add a little parser with a version check
```py
if __name__ == "__main__":
    assert pyro.__version__.startswith('1.3.1')
    parser = argparse.ArgumentParser(description="Toy mixture model")
    parser.add_argument("-n", "--num-steps", default=4000, type=int)
    args = parser.parse_args()
    main(args)
```
Then add plumbing of `args` through to `train()`. Then in `test_examples.py` add `--num-samples=1` to the invocation (as I've commented below).

Comment (garrett-bernstein) [reply examples/toy_mixture_model.py:40]: Done.

Comment (fritzo) [reply examples/toy_mixture_model.py:40]: Great, can you remove this `smoke_test = ...` line since it is no longer used?

Comment (garrett-bernstein) [reply examples/toy_mixture_model.py]: Whoops, done.

**Comment by martinjankowiak** [comment]
thanks for contributing @garrett-bernstein !

**Comment by garrett-bernstein** [comment]
Happy to help, thanks for the great package!

**Comment by garrett-bernstein** [comment]
@martinjankowiak Is it too late to easily change the heading that'll appear on the documentation sidebar to something like "Toy Mixture Model With Discrete Enumeration"? I realized the point of the script is to provide a succinct example for discrete enumeration, so it'll be a lot harder for people to find if they don't think to look for mixture models.

**Comment by martinjankowiak** [comment]
@garrett-bernstein sure just submit a follow-up PR
